association.go

```diff
@@ -1537,15 +1662,81 @@ func (a *Association) handleSack(d *chunkSelectiveAck) error {
 		a.awakeWriteLoop()
 	}
 
-	if a.inflightQueue.size() > 0 {
+	a.postprocessSack(state, cumTSNAckPointAdvanced)
+
+	return nil
+}
+
+// The caller must hold the lock. This method was only added because the
+// linter was complaining about the "cognitive complexity" of handleSack.
+func (a *Association) postprocessSack(state uint32, shouldAwakeWriteLoop bool) {
+	switch {
+	case a.inflightQueue.size() > 0:
 		// Start timer. (noop if already started)
 		a.log.Tracef("[%s] T3-rtx timer start (pt3)", a.name)
 		a.t3RTX.start(a.rtoMgr.getRTO())
+	case state == shutdownPending:
+		// No more outstanding, send shutdown.
+		shouldAwakeWriteLoop = true
+		a.willSendShutdown = true
+		a.setState(shutdownSent)
+	case state == shutdownReceived:
+		// No more outstanding, send shutdown ack.
+		shouldAwakeWriteLoop = true
+		a.willSendShutdownAck = true
+		a.setState(shutdownAckSent)
+	}
+
+	if shouldAwakeWriteLoop {
+		a.awakeWriteLoop()
```

Comment:
Hmm, I don't think I changed this behavior. The `shouldAwakeWriteLoop` argument in `sctp.Association.postprocessSack` is taken from the value of `cumTSNAckPointAdvanced` at [association.go:1665](https://github.com/pion/sctp/blob/85af7427b4a18641424f1b119bfd675b3e500d36/association.go#L1665) at the end of the `sctp.Association.handleSack` method.

So previously we had:

```go	
func (a *Association) handleSack(d *chunkSelectiveAck) error {
	// ... 

	if a.inflightQueue.size() > 0 {
		// Start timer. (noop if already started)
		a.log.Tracef("[%s] T3-rtx timer start (pt3)", a.name)
		a.t3RTX.start(a.rtoMgr.getRTO())
	}

	if cumTSNAckPointAdvanced {
		a.awakeWriteLoop()
	}

	return nil
}
```

And now we have:

```go
func (a *Association) handleSack(d *chunkSelectiveAck) error {
	// ... 

	a.postprocessSack(state, cumTSNAckPointAdvanced)

	return nil
}

// The caller must hold the lock. This method was only added because the
// linter was complaining about the "cognitive complexity" of handleSack.
func (a *Association) postprocessSack(state uint32, shouldAwakeWriteLoop bool) {
	switch {
	case a.inflightQueue.size() > 0:
		// Start timer. (noop if already started)
		a.log.Tracef("[%s] T3-rtx timer start (pt3)", a.name)
		a.t3RTX.start(a.rtoMgr.getRTO())
	case state == shutdownPending:
		// No more outstanding, send shutdown.
		shouldAwakeWriteLoop = true
		a.willSendShutdown = true
		a.setState(shutdownSent)
	case state == shutdownReceived:
		// No more outstanding, send shutdown ack.
		shouldAwakeWriteLoop = true
		a.willSendShutdownAck = true
		a.setState(shutdownAckSent)
	}

	if shouldAwakeWriteLoop {
		a.awakeWriteLoop()
	}
}
```

If you still have any concerns, it's possible that I missed something here.

Reply:
Oh I see, my bad.  It looks good to me now.